astartes/samplers/extrapolation/scaffold.py

```diff
@@ -1,5 +1,109 @@
+"""
+This sampler partitions the data based on the Bemis-Murcko scaffold function as implemented in RDKit.
+Bemis, G. W.; Murcko, M. A. The Properties of Known Drugs. 1. Molecular Frameworks. J. Med. Chem. 1996, 39, 2887−2893.
+Landrum, G. et al. RDKit: Open-Source Cheminformatics; 2006; https://www.rdkit.org.
+
+The goal is to cluster molecules that share the same scaffold.
+Later, these clusters will be assigned to training, validation, and testing split
+to create data splits that will measure extrapolation by testing on scaffolds
+that are not in the training set.
+
+"""
+import warnings
+from collections import defaultdict
+
+import numpy as np
+import rdkit
+from rdkit import Chem
+from rdkit.Chem.Scaffolds import MurckoScaffold
+
 from astartes.samplers import AbstractSampler
+from astartes.utils.warnings import NoMatchingScaffold
 
 
 class Scaffold(AbstractSampler):
-    pass
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        if not isinstance(self.X[0], str) and not isinstance(
```

Comment:
I also realized, while implementing SPXY, that the way we have this entire setup done means that this LOC will not be executed (the super constructor exits, basically, meaning we never return here). This line needs to be moved _before_ the call to the super constructor, and a unit test needs to be added to make sure this executed (probably actually two unit tests, one for the left case and one for the right case (or should there only be one case?))

Reply:
This comment is right for the wrong reasons, we do execute the remainder of the Scaffold constructor but the error is raised during execution of the superclass constructor